deps/childprocess.lua

```diff
@@ -164,4 +165,144 @@ local function spawn(command, args, options)
+  if not pcall(function() isWindows = require('luvipath').isWindows end) then
+    isWindows = los.type() == 'win32'
+  end
+
+  local stringArgs = command..table.concat(arguments, ' ')
+  if isWindows then
+    file = 'cmd.exe'
+    args = {'/s', '/c', '"'..stringArgs..'"' }
+  else
+    file = '/bin/sh'
+    args = {'-c', stringArgs }
+  end
+
+  if options and options.shell then file = options.shell end
+
+  return {
+    cmd = command,
+    file = file,
+    args = args or {},
+    options = options or {},
+    callback = callback
+  }
+end
+
+local function execFile(file, args, options, callback)
+  -- Make callback, args and options optional
+  if type(args) == 'function' then
+    callback = args
+    args, options = {}, {}
+  elseif type(options) == 'function' then
+    callback = options
+    options = {}
+  elseif not args and not options and not callback then
+    callback = function() end
+    options = {}
+    args = {}
+  end
+
+  local opts = {
+    timeout = 0,
+    maxBuffer = 4 * 1024,
+    signal = 'SIGTERM'
+  }
+  table.foreach(opts, function(k, v)
+    if not options[k] then options[k] = v end
+  end)
+
+  local child = spawn(file, args, options)
+
+  local stdout, stderr = {}, {}
+  local exited, killed = false, false
+  local stdoutLen, stderrLen = 0, 0
+  local timeoutId
+  local err = {}
+  local called = 2
+
+  local function exitHandler(code, signal)
+    if timeoutId then
+      timer.clearTimeout(timeoutId)
+      timeoutId = nil
+    end
+    if exited then return end
+    called = called - 1
+    if called == 0 then
+      if signal then err.signal = signal end
+      if not code then
+        err.message = 'Command failed: '..file
+      elseif code == 0 then
+        err = nil
+      else
+        err.code = code
+        err.message = 'Command killed'
+      end
+      exited = true
+      if not callback then return end
+      callback(err, table.concat(stdout, ""), table.concat(stderr, ""))
+    end
+  end
+  local function onClose(_exitCode)
+    exitHandler(_exitCode, nil)
+  end
+  local function kill()
+    child.stdout:emit('close', 1, options.signal)
+    child.stderr:emit('close', 1, options.signal)
+    child:emit('close', 1, options.signal)
+    killed = true
+    exitHandler(1, options.signal)
+  end
+
+  if options.timeout > 0 then
+    timeoutId = timer.setTimeout(options.timeout, function()
+      kill()
+      timeoutId = nil
+    end)
+  end
+
+  child.stdout:on('data', function(chunk)
+    stdoutLen = stdoutLen + #chunk
+    if stdoutLen > options.maxBuffer then
+      kill()
+    else
+      table.insert(stdout, chunk)
+    end
+  end):once('end', exitHandler)
+
+  child.stderr:on('data', function(chunk)
+    stderrLen = stderrLen + #chunk
+    if stderrLen > options.maxBuffer then
+      kill()
+    else
+      table.insert(stderr, chunk)
+    end
+  end)
+
+  child:once('close', onClose)
+end
+
+local function exec(command, options, callback)
+  -- options is optional
+  local opts
```

Comment:
If you modify `normalizeExecArgs` to return multiple values, this entire function can be simplified to:

```lua
return execFile(normalizeExecArgs(command, callback and options, callback or options)
```

`callback and options` will always be `options` if callback is true or falsy if callback is falsy.